be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala

```diff
@@ -1,86 +1,170 @@
-        handlerPartitioner.out(portPipelineError) ~> handlerMerger
-        handlerPartitioner.out(portResponseHandler) ~> responseHandler ~> handlerMerger
+          /* partitioner port numbers */
+          val portPipelineError = 0
+          val portResponseHandler = 1
+          val totalPorts = 2
 
-        /* close the shape */
-        FlowShape(handlerPartitioner.in, handlerMerger.out)
-      }
-  })
+          /* building blocks */
+          val handlerPartitioner = builder.add(Partition[GraphMessage](
+            totalPorts,
+            {
+              case Right(jsonRpcMessage: JsonRpcResponse) => jsonRpcMessage.result match {
+                  case Some(result) => result.resultMap match {
+                      case Some(_) => portResponseHandler
+                      case _       => portPipelineError
+                    }
+                  case _ => portPipelineError
+                }
+              case _ => portPipelineError // Pipeline error goes directly in handlerMerger
+            }
+          ))
+
+          val responseHandler = builder.add(GetMessagesByIdResponseHandler.responseHandler(mediatorActorRef, messageRegistry)(system))
+          val handlerMerger = builder.add(Merge[GraphMessage](totalPorts))
+
+          /* glue the components together */
+          handlerPartitioner.out(portPipelineError) ~> handlerMerger
+          handlerPartitioner.out(portResponseHandler) ~> responseHandler ~> handlerMerger
 
-  private def responseHandler(dbActorRef: ActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
+          /* close the shape */
+          FlowShape(handlerPartitioner.in, handlerMerger.out)
+        }
+    })
+
+  // This function packs each message into a publish before pushing into the pipeline
+  private def responseHandler(mediatorActorRef: ActorRef, messageRegistry: MessageRegistry)(implicit system: ActorSystem): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
     case Right(jsonRpcMessage: JsonRpcResponse) =>
-      val receivedResponse = jsonRpcMessage.result.get.resultMap.get
-      receivedResponse.keys.foreach(channel => {
-        receivedResponse(channel).foreach(message => {
-          writeOnDb(channel, message, dbActorRef, MAX_ATTEMPT)
-        })
-      })
-      Right(jsonRpcMessage)
+      val validatorFlow = validator(mediatorActorRef, messageRegistry)
+      val receivedResponse: Map[Channel, Set[GraphMessage]] =
+        jsonRpcMessage.result.get
+          .resultMap.get
+          .map {
+            case (channel, set) => (channel, set.map(msg => Right(JsonRpcRequest(RpcValidator.JSON_RPC_VERSION, MethodType.PUBLISH, new Publish(channel, msg), Some(0)))))
+          }
+
+      val failedMessages = passThroughPipeline(receivedResponse, validatorFlow, MAX_RETRY)
+      if (failedMessages.values.forall(set => set.isEmpty)) {
+        Right(JsonRpcResponse(RpcValidator.JSON_RPC_VERSION, new ResultObject(0), None))
```

Comment:
You could add a constant in result object such as `ResultObject SUCCESS` instead of the magic number